src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -2454,22 +2454,24 @@ void ShenandoahHeap::set_evacuation_reserve_quantities(bool is_valid) {
 }
 
 void ShenandoahHeap::set_concurrent_young_mark_in_progress(bool in_progress) {
-  if (has_forwarded_objects()) {
-    set_gc_state_mask(YOUNG_MARKING | UPDATEREFS, in_progress);
-  } else {
+  assert(!has_forwarded_objects(), "Young marking is not concurrent with evacuation");
+  if (!in_progress && is_concurrent_old_mark_in_progress()) {
+    assert(mode()->is_generational(), "Only generational GC has old marking");
+    // If old-marking is in progress when we turn off YOUNG_MARKING, leave MARKING (and OLD_MARKING) on
```

Comment:
Can we assert that gc state mask has `MARKING` set here?

Reply:
adding this

Reply:
pursuing this, but not adding redundant call to is_generational() along the non-ASSERT control path.

Reply:
Actually, this is too strong.  The suggested assert fails because we may already be doing OLD_MARKING when we start to do the YOUNG_MARKING.